macros/src/bitflags.rs

```diff
@@ -0,0 +1,153 @@
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{quote, ToTokens};
+use syn::{
+    parse::{Parse, ParseStream},
+    parse_macro_input,
+    punctuated::Punctuated,
+    token, Attribute, Expr, Ident, Token, Type, Visibility,
+};
+
+use crate::{mksym, symbol::Symbol};
+
+pub(super) fn expand(ts: TokenStream) -> TokenStream {
+    let ts2 = proc_macro2::TokenStream::from(ts.clone());
+    let input: Input = parse_macro_input!(ts);
+
+    let codegen = Codegen::new(&input);
+
+    // Encode package and disambiguator to provide the decoder with all info it needs (even if
+    // technically redundant, since it's also stored in the symbol we create).
+    let format_str = format!(
+        "{{={}:__internal_bitflags_{}@{}@{}}}",
+        input.ty.to_token_stream(),
+        input.ident,
+        crate::symbol::package(),
+        crate::symbol::disambiguator(),
+    );
+    let sym = mksym(&format_str, "bitflags", false);
+
+    let ident = &input.ident;
+    let ty = input.ty;
+    let flags = codegen.flags.iter().map(|f| &f.def);
+    let result = quote! {
+        const _: () = {
+            fn assert<T: defmt::export::UnsignedInt>() {}
+            assert::<#ty>;
+
+            #(#flags)*
+        };
+
+        defmt::export::bitflags! {
+            #ts2
+        }
+
+        impl defmt::Format for #ident {
+            fn format(&self, f: defmt::Formatter) {
+                unreachable!()
+            }
+            fn _format_tag() -> defmt::Str {
+                #sym
+            }
+            fn _format_data(&self) {
+                // There's a method available for every supported bitflags type.
+                defmt::export::#ty(&self.bits());
+            }
+        }
+    };
+    result.into()
+}
+
+struct FlagStatic {
+    def: TokenStream2,
+}
+
+struct Codegen {
+    flags: Vec<FlagStatic>,
+}
+
+impl Codegen {
+    fn new(input: &Input) -> Self {
+        let flags = input
+            .flags
+            .iter()
+            .map(|flag| {
+                let name = &flag.ident;
+                let value = &flag.value;
+                let repr_ty = &input.ty;
+
+                let sym = Symbol::new(
+                    "bitflags_value",
+                    &format!("{}::{}", input.ident, flag.ident),
+                )
+                .mangle();
+
+                let def = quote! {
+                    #[cfg_attr(target_os = "macos", link_section = ".defmt,end")]
+                    #[cfg_attr(not(target_os = "macos"), link_section = ".defmt.end")]
+                    #[export_name = #sym]
+                    static #name: u128 = {
```

Comment:
wait, I'm confused about this. why is this `u128`? this static lives in the fake "defmt" address space so we should *not* be loading (LDR instructions) values that have "defmt" addresses. also, we want to keep those address apart from each other by a single byte, for compaction purposes (this is a 16-bit address space)

is the target using the `coerced_value`s?.

Reply:
The target just uses what the `bitflags` crate generates, I've chosen to cast the value to a `u128` here so that we can treat all underlying types the same (the decoder will store all unsigned integers in a `u128` already, so this is convenient).

Reply:
Ah, and to expand on this a bit: the address of these statics is irrelevant, they are never referenced by the program, and they don't contribute to the `.defmt` address space usage (in 9ad68bb, I've changed how we compute how much of that address space is in use to make them not count towards the limit).

All these statics end up in `.defmt.end`, which is just a metadata section (and should maybe be renamed to reflect that). We're only interested in their symbol names and value.